### PR TITLE
Compile CSFML with SFML 3.0

### PR DESCRIPTION
**Pull request by Marioalexsan**
Changes a lot of places in code to compile under SFML 3.0.
Currently doesn't compile for MacOS due to `<uchar.h>` not being a thing for whatever reason.

### CODE REVIEW

File: include/SFML/Uchar.h [hunk 1]
```diff
@@ -22,22 +22,21 @@
...
-};
+#include <SFML/Config.h>
 
+#ifndef CSFML_SYSTEM_MACOS
+    #include <uchar.h>
+    typedef char32_t sfChar32;
+#else
+    #include <stdint.h>
+    typedef uint_least32_t sfChar32;
+#endif
```
Comment (ChrisThrasher) [inline include/SFML/Uchar.h:40]: Let's put this stuff in Config.h. That's where the old typedefs used to live before I recently removed them.

File: src/SFML/Network/Ftp.cpp [hunk 2]
```diff
@@ -115,10 +117,24 @@ const char* sfFtpDirectoryResponse_getMessage(const sfFtpDirectoryResponse* ftpD
 
 ////////////////////////////////////////////////////////////
 const char* sfFtpDirectoryResponse_getDirectory(const sfFtpDirectoryResponse* ftpDirectoryResponse)
+{
+    CSFML_CHECK_RETURN(ftpDirectoryResponse, nullptr);
+    return strdup(sf::String(ftpDirectoryResponse->This.getDirectory()).toAnsiString().c_str());
```
Comment (ChrisThrasher) [inline src/SFML/Network/Ftp.cpp:122]:
```suggestion
    return strdup(sf::String(ftpDirectoryResponse->This.getDirectory()).toAnsiString().c_str());
    return strdup(ftpDirectoryResponse->This.getDirectory().string().c_str());
```

No need to use sf::String here. `getDirector()` returns a `std::filesystem::path` which has a `.string()` function that returns an ASCII string from which we can get a `const char*`.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]

File: include/SFML/Uchar.h (hunk 1 above)
Comment (Marioalexsan) [reply include/SFML/Uchar.h]: Moved to config, let me know if any more changes are needed.

File: include/SFML/Config.h [hunk 3]
```diff
@@ -142,4 +142,12 @@
 
 #endif
 
+
+////////////////////////////////////////////////////////////
+// Define sfChar32
+////////////////////////////////////////////////////////////
```
Comment (ChrisThrasher) [inline include/SFML/Config.h:148]:
```suggestion
////////////////////////////////////////////////////////////
/// Define sfChar32
///
/// Work around macOS not providing <uchar.h>
////////////////////////////////////////////////////////////
```

File: include/SFML/Graphics/RenderWindow.h [hunk 4]
```diff
@@ -52,7 +52,7 @@
 /// \param settings Creation settings (pass NULL to use default values)
 ///
 ////////////////////////////////////////////////////////////
-CSFML_GRAPHICS_API sfRenderWindow* sfRenderWindow_create(sfVideoMode mode, const char* title, uint32_t style, const sfContextSettings* settings);
+CSFML_GRAPHICS_API sfRenderWindow* sfRenderWindow_create(sfVideoMode mode, const char* title, uint32_t style, uint32_t state, const sfContextSettings* settings);
```
Comment (ChrisThrasher) [inline include/SFML/Graphics/RenderWindow.h:55]: Can we use the `state` enum type here (and elsewhere were states are used) instead of `uint32_t`?

File: include/SFML/Window/WindowBase.h [hunk 5]
```diff
@@ -47,11 +47,21 @@ typedef enum
...
+////////////////////////////////////////////////////////////
+/// \brief Enumeration of the window states
+///
+////////////////////////////////////////////////////////////
+typedef enum
+{
+    sfWindowed,  //!< Floating window
+    sfFullscreen //!< Fullscreen window
+} sfWindowState;
+
```
Comment (ChrisThrasher) [inline include/SFML/Window/WindowBase.h:64]:
```suggestion


```

File: src/SFML/Char32.hpp [hunk 6]
```diff
@@ -22,27 +22,49 @@
 //
 ////////////////////////////////////////////////////////////
 
-#ifndef SFML_THREADSTRUCT_H
-#define SFML_THREADSTRUCT_H
+#ifndef SFML_CHAR32_H
+#define SFML_CHAR32_H
```
Comment (ChrisThrasher) [inline src/SFML/Char32.hpp:26]:
```suggestion
#ifndef SFML_CHAR32_HPP
#define SFML_CHAR32_HPP
```

File: src/SFML/Char32.hpp [hunk 7]
```diff
@@ -22,27 +22,49 @@
...
+
+    std::size_t byteCount = sizeof(sfChar32) * length;
+
+    memcpy(str.data(), utf32, byteCount);
 
+    return sf::String(str.data());
+}
 
-#endif // SFML_THREADSTRUCT_H
+#endif // SFML_CHAR32_H
```
Comment (ChrisThrasher) [inline src/SFML/Char32.hpp:70]:
```suggestion
#endif // SFML_CHAR32_HPP

```
Missing ending newline too

File: src/SFML/Char32.hpp [hunk 8]
```diff
@@ -22,27 +22,49 @@
...
 
-    sf::Thread This;
-};
+    std::u32string str(length, ' ');
+
+    std::size_t byteCount = sizeof(sfChar32) * length;
+
+    memcpy(str.data(), utf32, byteCount);
 
+    return sf::String(str.data());
```
Comment (ChrisThrasher) [inline src/SFML/Char32.hpp:67]: Nit: No need to double-space all the code. You can remove most of these empty newlines.

File: src/SFML/Graphics/Rect.cpp [hunk 9]
```diff
@@ -68,49 +68,51 @@ bool sfFloatRect_intersects(const sfFloatRect* rect1, const sfFloatRect* rect2,
...
     {
-        sf::FloatRect overlap;
-        bool intersects = SFMLRect1.intersects(SFMLRect2, overlap);
+        std::optional<sf::FloatRect> overlap = SFMLRect1.findIntersection(SFMLRect2);
 
-        intersection->left   = overlap.left;
-        intersection->top    = overlap.top;
-        intersection->width  = overlap.width;
-        intersection->height = overlap.height;
+        if (overlap) {
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Rect.cpp:78]:
```suggestion
        if (overlap)
        {
```

File: src/SFML/Graphics/Rect.cpp [hunk 10]
```diff
@@ -68,49 +68,51 @@ bool sfFloatRect_intersects(const sfFloatRect* rect1, const sfFloatRect* rect2,
...
     {
-        sf::IntRect overlap;
-        bool intersects = SFMLRect1.intersects(SFMLRect2, overlap);
+        std::optional<sf::IntRect> overlap = SFMLRect1.findIntersection(SFMLRect2);
 
-        intersection->left   = overlap.left;
-        intersection->top    = overlap.top;
-        intersection->width  = overlap.width;
-        intersection->height = overlap.height;
+        if (overlap) {
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Rect.cpp:104]:
```suggestion
        if (overlap)
        {
```

File: src/SFML/Graphics/RenderTexture.cpp [hunk 11]
```diff
@@ -40,19 +40,6 @@
...
-sfRenderTexture* sfRenderTexture_create(unsigned int width, unsigned int height, bool depthBuffer)
-{
-    sfRenderTexture* renderTexture = new sfRenderTexture;
-    renderTexture->This.create(width, height, depthBuffer);
-    renderTexture->Target = new sfTexture(const_cast<sf::Texture*>(&renderTexture->This.getTexture()));
-    renderTexture->DefaultView.This = renderTexture->This.getDefaultView();
-    renderTexture->CurrentView.This = renderTexture->This.getView();
-
-    return renderTexture;
-}
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/RenderTexture.cpp:53]: Why was this removed?

File: src/SFML/Graphics/Text.cpp [hunk 12]
```diff
@@ -31,15 +31,15 @@
...
 ////////////////////////////////////////////////////////////
-sfText* sfText_create(void)
+sfText* sfText_create(const sfFont* font)
 {
-    sfText* text = new sfText;
-    text->Font = nullptr;
+    CSFML_CHECK_RETURN(font, nullptr);
 
-    return text;
+    return new sfText{ sf::Text{font->This}, font, {}, {}, {} };
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Text.cpp:42]: I like how this new pattern ensures all fields get initialized. I believe the previous code left certain things uninitialized.

File: src/SFML/Network/Ftp.cpp [hunk 13]
```diff
@@ -115,10 +117,24 @@ const char* sfFtpDirectoryResponse_getMessage(const sfFtpDirectoryResponse* ftpD
...
+const sfChar32* sfFtpDirectoryResponse_getDirectoryUnicode(const sfFtpDirectoryResponse* ftpDirectoryResponse)
 {
     CSFML_CHECK_RETURN(ftpDirectoryResponse, nullptr);
 
-    return ftpDirectoryResponse->This.getDirectory().c_str();
+    std::u32string str = sf::String(ftpDirectoryResponse->This.getDirectory().c_str()).toUtf32();
+
+    std::size_t size = sizeof(sfChar32) * str.length();
+    sfChar32* ret = static_cast<sfChar32*>(malloc(size));
+    memcpy(ret, str.c_str(), size);
```
Comment (ChrisThrasher) [inline src/SFML/Network/Ftp.cpp:135]: Could this use the new conversion functions you wrote?

File: src/SFML/Network/TcpSocket.cpp [hunk 14]
```diff
@@ -74,8 +74,11 @@ sfIpAddress sfTcpSocket_getRemoteAddress(const sfTcpSocket* socket)
     sfIpAddress result = sfIpAddress_None;
     CSFML_CHECK_RETURN(socket, result);
 
-    sf::IpAddress address = socket->This.getRemoteAddress();
-    strncpy(result.address, address.toString().c_str(), 15);
+    std::optional<sf::IpAddress> address = socket->This.getRemoteAddress();
+
+    if (address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/TcpSocket.cpp:79]:
```suggestion
    if (address)
    {
```

File: src/SFML/Network/IpAddress.cpp [hunk 15]
```diff
@@ -33,18 +33,21 @@
 namespace
 {
     // Helper function for converting a SFML address to a CSFML one
-    sfIpAddress fromSFMLAddress(sf::IpAddress address)
+    sfIpAddress fromSFMLAddress(std::optional<sf::IpAddress> address)
     {
         sfIpAddress result = {0};
-        strncpy(result.address, address.toString().c_str(), 15);
+
+        if (address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/IpAddress.cpp:40]:
```suggestion
        if (address)
        {
```

File: src/SFML/Network/TcpSocket.cpp [hunk 16]
```diff
@@ -91,11 +94,15 @@ unsigned short sfTcpSocket_getRemotePort(const sfTcpSocket* socket)
 ////////////////////////////////////////////////////////////
 sfSocketStatus sfTcpSocket_connect(sfTcpSocket* socket, sfIpAddress remoteAddress, unsigned short remotePort, sfTime timeout)
 {
-    sf::IpAddress address(remoteAddress.address);
+    std::optional<sf::IpAddress> address = sf::IpAddress::resolve(remoteAddress.address);
+
+    if (!address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/TcpSocket.cpp:99]:
```suggestion
    if (!address)
    {
```

File: src/SFML/Network/UdpSocket.cpp [hunk 17]
```diff
@@ -72,8 +72,13 @@ sfSocketStatus sfUdpSocket_bind(sfUdpSocket* socket, unsigned short port, sfIpAd
 {
     CSFML_CHECK_RETURN(socket, sfSocketError);
 
-    sf::IpAddress sfmlAddress(address.address);
-    return static_cast<sfSocketStatus>(socket->This.bind(port, sfmlAddress));
+    std::optional<sf::IpAddress> sfmlAddress = sf::IpAddress::resolve(address.address);
+
+    if (!sfmlAddress) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/UdpSocket.cpp:77]:
```suggestion
    if (!sfmlAddress)
    {
```

File: src/SFML/Network/UdpSocket.cpp [hunk 18]
```diff
@@ -90,9 +95,13 @@ sfSocketStatus sfUdpSocket_send(sfUdpSocket* socket, const void* data, size_t si
     CSFML_CHECK_RETURN(socket, sfSocketError);
 
     // Convert the address
-    sf::IpAddress address(remoteAddress.address);
+    std::optional<sf::IpAddress> address = sf::IpAddress::resolve(remoteAddress.address);
+
+    if (!address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/UdpSocket.cpp:100]:
```suggestion
    if (!address)
    {
```

File: src/SFML/Network/UdpSocket.cpp [hunk 19]
```diff
@@ -115,7 +124,10 @@ sfSocketStatus sfUdpSocket_receive(sfUdpSocket* socket, void* data, size_t size,
     if (remoteAddress)
     {
         *remoteAddress = sfIpAddress_None;
-        strncpy(remoteAddress->address, address.toString().c_str(), 15);
+
+        if (address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/UdpSocket.cpp:128]:
```suggestion
        if (address)
        {
```

File: src/SFML/Network/UdpSocket.cpp [hunk 20]
```diff
@@ -132,9 +144,13 @@ sfSocketStatus sfUdpSocket_sendPacket(sfUdpSocket* socket, sfPacket* packet, sfI
     CSFML_CHECK_RETURN(packet, sfSocketError);
 
     // Convert the address
-    sf::IpAddress address(remoteAddress.address);
+    std::optional<sf::IpAddress> address = sf::IpAddress::resolve(remoteAddress.address);
+
+    if (!address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/UdpSocket.cpp:149]:
```suggestion
    if (!address)
    {
```

File: src/SFML/Network/UdpSocket.cpp [hunk 21]
```diff
@@ -144,17 +160,20 @@ sfSocketStatus sfUdpSocket_receivePacket(sfUdpSocket* socket, sfPacket* packet,
...
-    if (status != sf::Socket::Done)
+    if (status != sf::Socket::Status::Done)
         return static_cast<sfSocketStatus>(status);
 
     if (remoteAddress)
     {
         *remoteAddress = sfIpAddress_None;
-        strncpy(remoteAddress->address, address.toString().c_str(), 15);
+
+        if (address) {
```
Comment (ChrisThrasher) [inline src/SFML/Network/UdpSocket.cpp:174]:
```suggestion
        if (address)
        {
```

File: src/SFML/System/BufferStruct.h [hunk 22]
```diff
@@ -37,7 +37,7 @@
 ////////////////////////////////////////////////////////////
 struct sfBuffer
 {
-    std::vector<sf::Uint8> buffer;
+    std::vector<uint8_t> buffer;
```
Comment (ChrisThrasher) [inline src/SFML/System/BufferStruct.h:40]:
```suggestion
    std::vector<std::uint8_t> buffer;
```

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]

File: src/SFML/Graphics/RenderTexture.cpp (hunk 11 above)
Comment (Marioalexsan) [reply src/SFML/Graphics/RenderTexture.cpp:53]: The function was deprecated according to RenderTexture.h

**Comment by ChrisThrasher** [comment]
This PR is shaping up well! We'll need a few more rounds of reviews but I think we can land this in the next few days.

File: src/SFML/Graphics/RenderTexture.cpp (hunk 11 above)
Comment (ChrisThrasher) [reply src/SFML/Graphics/RenderTexture.cpp:53]: Ah okay thanks

File: src/SFML/Network/Ftp.cpp (hunk 13 above)
Comment (Marioalexsan) [reply src/SFML/Network/Ftp.cpp]: Done.

File: include/SFML/Graphics/RenderWindow.h (hunk 4 above)
Comment (Marioalexsan) [reply include/SFML/Graphics/RenderWindow.h]: Should be done.

**Review by Jim-Marsden — APPROVED** [review]

File: include/SFML/Graphics/RenderTexture.h [hunk 23]
```diff
@@ -40,21 +40,6 @@
...
-/// \param height      Height of the render texture
-/// \param depthBuffer Do you want a depth-buffer attached? (useful only if you're doing 3D OpenGL on the rendertexture)
-///
-/// \return A new sfRenderTexture object, or NULL if it failed
-///
-/// \deprecated
-/// Use sfRenderTexture_createWithSettings instead.
-///
-////////////////////////////////////////////////////////////
-CSFML_GRAPHICS_API sfRenderTexture* sfRenderTexture_create(unsigned int width, unsigned int height, bool depthBuffer);
```
[the remark this answers is not in the document]
Comment (Marioalexsan) [reply include/SFML/Graphics/RenderTexture.h:56]: Sounds logical, I went ahead and did it.

File: src/SFML/Graphics/Image.cpp [hunk 24]
```diff
@@ -137,7 +137,17 @@ bool sfImage_saveToFile(const sfImage* image, const char* filename)
 bool sfImage_saveToMemory(const sfImage* image, sfBuffer* output, const char* format)
 {
     CSFML_CHECK_RETURN(output, false);
-    CSFML_CALL_RETURN(image, saveToMemory(output->buffer, format), false);
+    CSFML_CHECK_RETURN(image, false);
+
+    std::optional<std::vector<std::uint8_t>> data = image->This.saveToMemory(format);
```
[the remark this answers is not in the document]
Comment (Marioalexsan) [reply src/SFML/Graphics/Image.cpp]: Done. Let me know if there's other spots that could use it.

**Review by texus — APPROVED** [review]
`copyFromChar32` in `src/Char32.hpp` isn't used anywhere anymore and can probably be removed.

**Review by ChrisThrasher — APPROVED** [review]
I'm happy with how far this PR has come. I think it's ready to merge.
